posts/bonus-fastify-encapsulation.md

```diff
@@ -0,0 +1,324 @@
+
+Here is a descriptive image:
+
+[![](https://mermaid.ink/img/pako:eNpFjz0PgjAQhv8KuRkGHTuY8LEZJx0U26GhpxBtS-qRaAj_3UMEbrr3uWe4t4fKGwQB96DbOjoV0kU86TX3jvBN0UYlyS5d81lNRraQ7Whka77MRsY8zxe-nznjslxw-cer6BTEYDFY3Rj-rB_vEqhGixIEr0aHhwTpBvZ0R_74cRUICh3G0LVGExaN5kIWxE0_X0zRNOTDYar6azx8ATfET_I)](https://mermaid.live/edit#pako:eNpFjz0PgjAQhv8KuRkGHTuY8LEZJx0U26GhpxBtS-qRaAj_3UMEbrr3uWe4t4fKGwQB96DbOjoV0kU86TX3jvBN0UYlyS5d81lNRraQ7Whka77MRsY8zxe-nznjslxw-cer6BTEYDFY3Rj-rB_vEqhGixIEr0aHhwTpBvZ0R_74cRUICh3G0LVGExaN5kIWxE0_X0zRNOTDYar6azx8ATfET_I)
+
+This new graph is more complex, but it is very powerful because it defines new behaviours:
+
+- Each context may have child contexts.
+- Each context could have at most one parent context.
+- Each context can access the state of its parent context.
+
+That being said, we have mentioned all the concepts we need to understand the Fastify encapsulation.
+Let's see how it works in Fastify.
+
+
+## The Plugin System demystified
+
+In one sentence: Fastify's Plugin System is an encapsulated context builder. No more, no less.  
+
+The Plugins System has **these pillars**:
+
+1. There is only one root context, and all the plugins are children of it
+2. The parent context can't access the state of the child contexts
+3. The children's context can access the state of the parent's context
+
+I'm used to calling the root context the _root application instance_ and every child context as _plugin instance_.  
+Since we have a root context, every Fastify application can be represented as a [tree structure](https://en.wikipedia.org/wiki/Tree_(data_structure) where every node is a context.
+
+Let's take a look at some code to understand better how it works.
+
+Create a new project:
+
+```bash
+mkdir fastify-encapsulation
+cd fastify-encapsulation
+npm init --yes
+npm install fastify
+touch app.js
+```
+
+In a new `app.js` file, add the following code:
+
+```js
+const fastify = require('fastify')
+
+start()
+
+async function start() {
+  const app = fastify({ logger: true })
+  await app.listen(8080)
+}
+```
+
+The `app` variable in the previous code is the _root application instance_ or Fastify's root context.
+
+Let's try to create some contexts, as shown in the previous image.  
+To do this, we need to call the [`register`](https://www.fastify.io/docs/latest/Reference/Plugins/) method.
+
+Edit the `start` function as follows:
+
+```js
+const app = fastify({ logger: true })
+
+app.register(async function plugin (instance, opts) {
+  // context 1
+  instance.register(async function plugin (instance, opts) {
+  // context X
+  })
+})
+
+app.register(async function plugin (instance, opts) {
+  // context 2
+  instance.register(async function plugin (instance, opts) {
+    // context Y
+    instance.register(async function plugin (instance, opts) {
+      // context K
+    })
+  })
+  instance.register(async function plugin (instance, opts) {
+    // context Z
+  })
+})
+```
```

Comment:
would it be worth giving each plugin a unique name? It might confuse some readers that they are each a function named `plugin`.